### PR TITLE
feat: orchestrator mint preparation, confirmation, and authorization validation

### PR DESCRIPTION
**Pull request by rouzwelt**
## Motivation

The orchestrator contract supports a `mint()` path that is distinct from vault-direct deposits: it asserts 1:1 share minting on-chain, keeps the receipt, and forwards ERC-20 shares directly to the recipient. The service layer had no support for preparing, confirming, or recovering orchestrator mints, and there was no validation path for the recipient-produced `MintAuthV1` authorization that the orchestrator requires before it will execute a mint.

Additionally, the mint-side revert reasons (`NonceReplayed`, `BadRecipientSignature`, `RecipientCallbackRejected`, `VaultAmountMismatch`) were absent from `OrchestratorRevertReason`, meaning a reverted mint transaction could not be classified and the burn failure classifier would panic on an incomplete match.

## Solution

Four new `VaultService` methods are introduced:

- `prepare_orchestrator_mint_tx` — builds and signs the `ST0xOrchestrator.mint(token, to, amount, auth, receiptInformation)` transaction without broadcasting it. The calldata carries exactly the signed `(token, to, amount, nonce)` from the authorization; there is no `previewDeposit` and no multicall.
- `confirm_orchestrator_mint` — waits for the transaction receipt, parses the `Minted` event, and on a mined-but-reverted transaction replays it as an `eth_call` to decode the typed revert reason.
- `find_orchestrator_minted_log` — queries `Minted` logs filtered by `(token, to)` and full-matches `(to, nonce, token, amount)`. A bare `(to, nonce)` hit whose token or amount differs is explicitly not a match, since the on-chain uniqueness key is only `(to, nonce)` and a partial hit means a different mint consumed the nonce.
- `validate_mint_authorization` — checks the consumed-nonce flag first, then for a non-empty signature recovers the signer via ECDSA (EOA) or ERC-1271 `isValidSignature` (contract recipient). An empty signature skips recovery entirely, covering the future bridge/`IMintRecipient` path where the recipient authorizes via the orchestrator's `authorizeMint` callback.

The `OrchestratorRevertReason` enum gains `NonceReplayed`, `BadRecipientSignature`, `RecipientCallbackRejected`, and `VaultAmountMismatch`. The burn failure classifier's `match` is updated to explicitly route all mint-side variants to `Unclassified`, with a comment explaining they cannot arise from a burn. Three new `VaultError` variants are added: `MintAuthSignerMismatch`, `MintAuthNonceUsed`, and `MintAuthRejectedByContract`. The `IERC1271` ABI binding is wired in for the ERC-1271 contract-recipient path.

The mock service is extended with full implementations of all four methods, configurable failure injection (`with_orchestrator_mint_confirm_revert`, `with_mint_auth_failure`, `with_minted_log`, `with_orchestrator_mint_result`), call counters, and `reset()` coverage for all new state.  
  
Closes [RAI-1615](https://linear.app/makeitrain/issue/RAI-1615/orchestrator-mint-preparation-confirmation-and-authorization)

## Checks

By submitting this for review, I'm confirming I've done the following:

- [x] added comprehensive test coverage for any changes in logic
- [x] made this PR as small as possible
- [x] linked any relevant issues or PRs

## Summary by CodeRabbit

- **New Features**
    - Added support for orchestrator-driven minting, including transaction preparation, confirmation, minted-log recovery, and tracking mint results (nonce, amount, shares, gas, block, tx identifiers).
    - Added recipient authorization verification via signatures, including ERC-1271 contract wallet support.
- **Bug Fixes**
    - Expanded typed revert decoding and failure classification for mint scenarios (replayed nonces, invalid recipient signatures, rejected callbacks, amount mismatches) and burn-side edge cases.
- **Tests**
    - Added extensive unit-test coverage for the full mint lifecycle and log/authorization matching semantics.

### CODE REVIEW

**Comment by rouzwelt** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.com/github/pr/ST0x-Technology/st0x.issuance/296?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#332** <a href="https://app.graphite.com/github/pr/ST0x-Technology/st0x.issuance/332?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 3  dependent PRs ([#293](https://github.com/ST0x-Technology/st0x.issuance/pull/293) <a href="https://app.graphite.com/github/pr/ST0x-Technology/st0x.issuance/293?utm_source=stack-comment-icon" target="_blank">Graphite</a>, [#306](https://github.com/ST0x-Technology/st0x.issuance/pull/306) <a href="https://app.graphite.com/github/pr/ST0x-Technology/st0x.issuance/306?utm_source=stack-comment-icon" target="_blank">Graphite</a>, [#335](https://github.com/ST0x-Technology/st0x.issuance/pull/335) <a href="https://app.graphite.com/github/pr/ST0x-Technology/st0x.issuance/335?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#331** <a href="https://app.graphite.com/github/pr/ST0x-Technology/st0x.issuance/331?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#317** <a href="https://app.graphite.com/github/pr/ST0x-Technology/st0x.issuance/317?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#324** <a href="https://app.graphite.com/github/pr/ST0x-Technology/st0x.issuance/324?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#309** <a href="https://app.graphite.com/github/pr/ST0x-Technology/st0x.issuance/309?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#308** <a href="https://app.graphite.com/github/pr/ST0x-Technology/st0x.issuance/308?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#303** <a href="https://app.graphite.com/github/pr/ST0x-Technology/st0x.issuance/303?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#302** <a href="https://app.graphite.com/github/pr/ST0x-Technology/st0x.issuance/302?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#301** <a href="https://app.graphite.com/github/pr/ST0x-Technology/st0x.issuance/301?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#300** <a href="https://app.graphite.com/github/pr/ST0x-Technology/st0x.issuance/300?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#298** <a href="https://app.graphite.com/github/pr/ST0x-Technology/st0x.issuance/298?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#296** <a href="https://app.graphite.com/github/pr/ST0x-Technology/st0x.issuance/296?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/ST0x-Technology/st0x.issuance/296?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#289** <a href="https://app.graphite.com/github/pr/ST0x-Technology/st0x.issuance/289?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#288** <a href="https://app.graphite.com/github/pr/ST0x-Technology/st0x.issuance/288?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#287** <a href="https://app.graphite.com/github/pr/ST0x-Technology/st0x.issuance/287?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#286** <a href="https://app.graphite.com/github/pr/ST0x-Technology/st0x.issuance/286?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#285** <a href="https://app.graphite.com/github/pr/ST0x-Technology/st0x.issuance/285?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#280** <a href="https://app.graphite.com/github/pr/ST0x-Technology/st0x.issuance/280?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#279** <a href="https://app.graphite.com/github/pr/ST0x-Technology/st0x.issuance/279?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#277** <a href="https://app.graphite.com/github/pr/ST0x-Technology/st0x.issuance/277?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#275** <a href="https://app.graphite.com/github/pr/ST0x-Technology/st0x.issuance/275?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#274** <a href="https://app.graphite.com/github/pr/ST0x-Technology/st0x.issuance/274?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#273** <a href="https://app.graphite.com/github/pr/ST0x-Technology/st0x.issuance/273?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#271** <a href="https://app.graphite.com/github/pr/ST0x-Technology/st0x.issuance/271?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#268** <a href="https://app.graphite.com/github/pr/ST0x-Technology/st0x.issuance/268?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#267** <a href="https://app.graphite.com/github/pr/ST0x-Technology/st0x.issuance/267?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#216** <a href="https://app.graphite.com/github/pr/ST0x-Technology/st0x.issuance/216?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add the label <kbd>add-to-gt-merge-queue</kbd> to this PR to add it to the merge queue.

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.com/invite/github/ST0x-Technology?ref=merge-queue-instructions-in-stack-comment&prId=296).

<h2></h2>
<sub>

An organization admin has required the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: src/vault/mod.rs [hunk 1]
```diff
@@ -262,6 +263,74 @@ pub(crate) trait VaultService: Send + Sync {
...
+        &self,
+        _orchestrator: Address,
+        _token: Address,
+        _to: Address,
+        _amount: U256,
+        _authorization: &MintAuthorization,
+    ) -> Result<(), VaultError> {
+        Err(VaultError::InvalidReceipt)
+    }
 }
```
[the remark this answers is not in the document]
Comment (rouzwelt) [reply src/vault/mod.rs:357]: its intentional and will be removed in upstack PRs

File: src/vault/service.rs [hunk 2]
```diff
@@ -863,6 +901,273 @@ impl VaultService for RealBlockchainService {
...
+                )
+                .await;
+            return Err(VaultError::OrchestratorReverted { tx_hash, reason });
+        }
+
+        let minted = receipt
+            .inner
+            .logs()
+            .iter()
+            .find_map(|log| {
```
Comment (0xgleb) [inline src/vault/service.rs:1189]: critical: this takes the first log in the receipt that decodes as `Minted` and lifts `nonce` and `amount` straight into the result, with no check that `log.address()` is the orchestrator, and it cannot cross-check `to`, `token` or `amount` because the method only receives `tx_id`. The mint transaction calls into the recipient at `to` before the mint completes, via ERC-1271 today and `authorizeMint` on the bridge path, so a contract at `to` emitting a log with the same topic0 and a larger amount gets picked up ahead of the orchestrator's real event, and we then record `OrchestratorTokensMinted` with shares that were never minted. `find_orchestrator_minted_log` below does this correctly, address-scoped and full-matching all four fields, so the primary success path validates strictly less than the recovery path for the same fact. Pass the expected mint facts in and share one helper between the two.

File: src/vault/mod.rs [hunk 3]
```diff
@@ -262,6 +263,74 @@ pub(crate) trait VaultService: Send + Sync {
...
+    /// `MintTxIntended`) before [`VaultService::submit_mint`], which
+    /// rebroadcasts them mode-agnostically. No `previewDeposit`, no
+    /// multicall — the orchestrator asserts 1:1 on-chain and forwards shares
+    /// to the recipient. The call uses exactly the signed
+    /// `(token, to, amount, nonce)` from the authorization.
+    ///
+    /// Production callers land with the orchestrator mint aggregate arms
+    /// (mirroring how `check_orchestrator_burn_readiness` above awaited its
+    /// caller); the allow is removed then.
+    #[allow(dead_code)]
```
Comment (0xgleb) [inline src/vault/mod.rs:300]: minor: this and the three sibling methods below each carry `#[allow(dead_code)]`, and a repo-wide grep at this commit shows none has a production caller, only `#[cfg(test)]` ones. AGENTS.md marks the lint policy CRITICAL and says never to add an `#[allow(...)]` without explicit permission. The justification in the comment, mirroring how `check_orchestrator_burn_readiness` awaited its caller, does not hold up: that method carries no suppression because it landed already wired into `burn_manager`.

Stack note: #300 removes all three suppressions once the methods have production callers, so the end state is clean and this is a property of the intermediate commit only. Raising it because the stated precedent is inaccurate rather than because the attribute survives.

File: src/vault/service.rs [hunk 4]
```diff
@@ -863,6 +901,273 @@ impl VaultService for RealBlockchainService {
...
+            return Err(VaultError::MintAuthNonceUsed {
+                to,
+                nonce: authorization.nonce,
+            });
+        }
+
+        // The bridge/`IMintRecipient` recipient authorizes on-chain via the
+        // orchestrator's `authorizeMint` callback — there is no signature to
+        // recover (SPEC Decision 1).
+        if authorization.signature.is_empty() {
```
Comment (0xgleb) [inline src/vault/service.rs:1358]: should fix: this returns `Ok` after only the consumed-nonce check, without confirming `to` actually has contract code. For a plain EOA recipient that authorization can never be satisfied on-chain, so it passes preflight and then reverts when the mint is broadcast, burning gas and pushing the failure into revert classification instead of catching it here. Gate this branch on `to` having code, using the same `get_code_at` call the branch below already makes.

File: src/vault/service.rs [hunk 5]
```diff
@@ -863,6 +901,273 @@ impl VaultService for RealBlockchainService {
...
+    async fn find_orchestrator_minted_log(
+        &self,
+        orchestrator: Address,
+        to: Address,
+        nonce: B256,
+        token: Address,
+        amount: U256,
+    ) -> Result<Option<OrchestratorMintedLog>, VaultError> {
+        let contract = IST0xOrchestratorV1::new(orchestrator, &self.provider);
+        let head = self.provider.get_block_number().await?;
```
Comment (0xgleb) [inline src/vault/service.rs]: should fix: this is the latest block, not a finalized or safe tag, and it becomes both the scan ceiling and the proof horizon. So a `Minted` log from a block that later reorgs out still counts as proof the mint landed, and recovery completes a mint whose shares no longer exist on chain. Query against a finality-tagged block, or require a confirmation depth before treating the log as proof.

File: src/vault/service.rs [hunk 6]
```diff
@@ -863,6 +901,273 @@ impl VaultService for RealBlockchainService {
...
+                .Minted_filter()
+                .topic2(token)
+                .topic3(to)
+                .from_block(from_block)
+                .to_block(to_block)
+                .query()
+                .await?;
+
+            for (minted, log) in logs {
+                if minted.nonce == nonce
```
Comment (0xgleb) [inline src/vault/service.rs]: should fix: when a log matches `(to, nonce)` but the token or amount differs, this condition simply falls through and the scan continues, so the caller cannot distinguish a near-miss from nothing found. On-chain uniqueness is only `(to, nonce)`, so a near-miss means a different mint already consumed that nonce, which is a materially different situation from the mint not having landed yet. Return a distinct error for that case so recovery escalates instead of retrying.

File: src/vault/mod.rs [hunk 7]
```diff
@@ -262,6 +263,74 @@ pub(crate) trait VaultService: Send + Sync {
...
+    }
+
+    /// Looks up a landed orchestrator mint by its `Minted` log,
+    /// full-matching `(to, nonce, token, amount)`. The on-chain uniqueness
+    /// key is only `(to, nonce)`, so a bare hit is not sufficient proof that
+    /// *this* mint landed — a hit whose token or amount differs means the
+    /// nonce was consumed by a different mint, and the caller must fail
+    /// manually rather than complete the wrong mint.
+    #[allow(dead_code)]
+    async fn find_orchestrator_minted_log(
```
Comment (0xgleb) [inline src/vault/mod.rs:330]: should fix: this takes `(orchestrator, to, nonce, token, amount)` while `validate_mint_authorization` below takes `token` before `to`. Both are `Address`, so transposing them at a call site compiles cleanly and silently queries or validates the wrong pair. Either align the order across the two, or wrap them in named types so the compiler catches it.

File: src/vault/service.rs [hunk 8]
```diff
@@ -863,6 +901,273 @@ impl VaultService for RealBlockchainService {
...
+        // orchestrator's `authorizeMint` callback — there is no signature to
+        // recover (SPEC Decision 1).
+        if authorization.signature.is_empty() {
+            return Ok(());
+        }
+
+        // The contract's own digest view guarantees the exact EIP-712 domain
+        // and struct hash the orchestrator will verify at mint time.
+        let digest = contract
+            .mintAuthDigest(token, to, amount, authorization.nonce)
```
Comment (0xgleb) [inline src/vault/service.rs]: should fix: this passes two same-typed `Address` values positionally, and nothing here pins that order to the contract's. If those two parameters are ever reordered on-chain the digest changes silently and every signature stops recovering to the expected signer, with no compile error anywhere. A test asserting the digest against a known-good fixture from the deployed contract would pin it.

File: src/vault/service.rs [hunk 9]
```diff
@@ -863,6 +901,273 @@ impl VaultService for RealBlockchainService {
...
+            .call()
+            .await?;
+
+        if self.provider.get_code_at(to).await?.is_empty() {
+            let signature =
+                Signature::from_raw(authorization.signature.as_ref())?;
```
Comment (0xgleb) [inline src/vault/service.rs:1372]: should fix: this fails on a wrong length or a bad `v`, and the diff adds `VaultError::Signature` for exactly that, but nothing drives it. These bytes arrive over the authorization channel, so malformed input is the expected case rather than an exotic one. Add a case with a truncated signature asserting the typed error.

File: src/vault/service.rs [hunk 10]
```diff
@@ -863,6 +901,273 @@ impl VaultService for RealBlockchainService {
...
+            let recovered = signature.recover_address_from_prehash(&digest)?;
```
Comment (0xgleb) [inline src/vault/service.rs:1373]: minor: this accepts a high-s signature that a Solidity `ecrecover` guard would reject, so a signature could pass preflight here and still fail on-chain. I have not checked what the orchestrator actually enforces, so this may well be a non-issue. Worth confirming against the contract, and normalizing s here if it does enforce low-s.

File: src/vault/service.rs [hunk 11]
```diff
@@ -863,6 +901,273 @@ impl VaultService for RealBlockchainService {
...
+        // transport failure.
+        match IERC1271::new(to, &self.provider)
+            .isValidSignature(digest, authorization.signature.clone())
+            .call()
+            .await
+        {
+            Ok(magic)
+                if magic
+                    == alloy::primitives::FixedBytes(
+                        IERC1271::isValidSignatureCall::SELECTOR,
```
Comment (0xgleb) [inline src/vault/service.rs:1398]: should fix: only this wrong-magic outcome is tested. The contract-recipient branch has two further failure outcomes below, a revert reinterpreted as rejection and a genuine transport error, and neither is driven by a test. The revert-as-rejection reinterpretation is the interesting one, since it deliberately converts an `Err` into `MintAuthRejectedByContract` and nothing currently pins that.

File: src/vault/service.rs [hunk 12]
```diff
@@ -863,6 +901,273 @@ impl VaultService for RealBlockchainService {
...
+                    ) =>
+            {
+                Ok(())
+            }
+            Ok(_) => Err(VaultError::MintAuthRejectedByContract { to }),
+            Err(error) if error.as_revert_data().is_some() => {
```
Comment (0xgleb) [inline src/vault/service.rs:1403]: minor: the comment says a contract that does not implement ERC-1271 is treated as a rejection, but `as_revert_data` returns `None` when a call succeeds with empty return data, which is exactly what a contract without that function does. That case falls through to the transport-error branch and propagates as an infrastructure failure rather than a rejection. Either handle empty return data explicitly or fix the comment.

File: src/vault/service.rs [hunk 13]
```diff
@@ -48,6 +51,20 @@ pub type RealBlockchainServiceProvider = FillProvider<
...
 >;
 
+/// Total backward window `find_orchestrator_minted_log` scans for a landed
+/// `Minted` log, from the chain head. A consuming transaction can only land
+/// after the mint's authorization was delivered (the delivery endpoint
+/// rejects an already-consumed nonce), so the sought log is at most as old
+/// as the mint's own recovery timeline — this window (~5.8 days on Base's 2s
+/// blocks) covers the automatic retry schedule and same-week manual
+/// re-drives with ample margin.
+const MINTED_LOG_LOOKBACK_BLOCKS: u64 = 250_000;
```
Comment (0xgleb) [inline src/vault/service.rs:62]: minor: this is justified as roughly 5.8 days on Base's 2 second blocks, but the constant applies to every configured network, so a slower chain gets a much longer window and a faster one a shorter one. A landing older than the window silently returns `None` and drops to the manual path. Worth deriving it from the chain config instead of hardcoding it.

File: src/vault/orchestrator.rs [hunk 14]
```diff
@@ -1,8 +1,94 @@
...
+/// `MintAuthorizationReceived`, so this serde shape is permanent.
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+pub(crate) struct MintAuthorization {
+    /// Recipient-chosen random nonce; `(to, nonce)` is single-use on-chain.
+    pub(crate) nonce: B256,
+    /// Opaque recipient signature over the `MintAuthV1` digest. EMPTY is a
+    /// valid input: the future Atomic Bridge recipient authorizes via the
+    /// orchestrator's `authorizeMint` callback instead of a signature, so
+    /// validation skips signer recovery for empty bytes (SPEC Decision 1).
+    pub(crate) signature: Bytes,
```
Comment (0xgleb) [inline src/vault/orchestrator.rs:24]: should fix: two things about this field. An empty value is a sentinel meaning the bridge recipient authorizes via the `authorizeMint` callback, so a second authorization mode is encoded as the absence of data, and nothing in the type stops an empty signature reaching the EOA path by accident. Separately, the struct has public fields and a derived `Deserialize`, so an unvalidated authorization is representable and persistable anywhere in the crate. An enum with one variant per mode, built through a fallible constructor, would make both invalid states unrepresentable, which matters more than usual here because this shape is meant to be permanent for a future event.

File: src/vault/mock.rs [hunk 15]
```diff
@@ -1513,23 +1655,236 @@ impl VaultService for MockVaultService {
...
+            // NOT a match — it is the "nonce consumed by a different mint"
+            // case.
+            return Ok(self
+                .orchestrator
+                .minted_log
+                .lock()
+                .unwrap()
+                .clone()
+                .filter(|log| {
+                    log.nonce == nonce && log.shares_minted == amount
```
Comment (0xgleb) [inline src/vault/mock.rs]: minor: this matches on nonce and amount only, while the real implementation full-matches nonce, token, `to` and amount. Upstack tests written against this mock will green-path a different-token or different-recipient collision that production correctly rejects. Tightening it here keeps those tests honest.

File: src/vault/mock.rs [hunk 16]
```diff
@@ -1969,4 +2324,250 @@ mod tests {
...
+            external_tx_id: None,
+        }
+    }
+
+    #[tokio::test]
+    async fn orchestrator_mint_mock_round_trip_records_and_confirms() {
+        let mock = MockVaultService::new_success();
+        let params = test_orchestrator_mint_params();
+
+        assert_eq!(mock.orchestrator_mint_preparation_call_count(), 0);
```
Comment (0xgleb) [inline src/vault/mock.rs:2467]: minor: this asserts a vault-direct parity for `InvalidPreparedMint` that does not exist on that path, so it documents a behaviour rather than verifying one. It will keep passing regardless of whether the parity is ever introduced, which makes it worse than no test here.

File: src/redemption/mod.rs [hunk 17]
```diff
@@ -391,7 +391,13 @@ const fn burn_failure_classification(
             OrchestratorRevertReason::ReceiptLogicMismatch => {
                 BurnFailureClassification::ReceiptLogicMismatch
             }
-            OrchestratorRevertReason::Unknown => {
+            // Mint-side reverts can never arise from a burn transaction;
+            // if one somehow surfaces here it stays unclassified.
+            OrchestratorRevertReason::NonceReplayed { .. }
+            | OrchestratorRevertReason::BadRecipientSignature
+            | OrchestratorRevertReason::RecipientCallbackRejected { .. }
+            | OrchestratorRevertReason::VaultAmountMismatch { .. }
```
Comment (0xgleb) [inline src/redemption/mod.rs:417]: minor: folding `VaultAmountMismatch` in with the other mint-side reasons rests entirely on the comment's claim that mint-side reverts can never arise from a burn. If that selector can ever surface on a burn it means the 1:1 vault ratio is broken, and `Unclassified` puts it on the auto-retry path rather than alerting. I could not settle whether the burn path can produce it, so worth confirming the selector really is mint-only.

File: tests/orchestrator_redemption.rs [hunk 18]
```diff
@@ -163,26 +163,19 @@ async fn wait_for_recovery_to_resolve(
...
-/// classifications are never auto-retried. Its persistent `/admin/stuck`
-/// entry proves the seeds replay and the stuck surfacing works, so a target
-/// redemption's absence in the same snapshot is positive evidence of
-/// recovery (see `wait_for_recovery_to_resolve`).
+/// Event stream for a control redemption parked in a classified
+/// `BurnFailed` — a state recovery provably never resolves (the reconciler
+/// skips typed classifications) — seeded alongside a recovery target so the
+/// absence-based assertion in [`wait_for_recovery_to_resolve`] is
+/// non-vacuous.
 fn classified_control_events(
```
Comment (0xgleb) [inline tests/orchestrator_redemption.rs:171]: minor: this is a burn-path test and the changes here are a refactor unrelated to the mint support the rest of the PR adds, including weakening this helper's return from a fixed-size array to a `Vec` when no caller needs the flexibility. It makes the diff larger than the change it describes. Worth splitting out, or at least reverting the array to `Vec` part.

**Review by 0xgleb — CHANGES_REQUESTED** [review]

File: src/redemption/mod.rs (hunk 17 above)
Comment (rouzwelt) [reply src/redemption/mod.rs:417]: Couldn't confirm from the ABI alone either, which is why the code already hedges ("if one somehow surfaces here it stays unclassified") rather than claiming impossibility. If it ever did surface on a burn, Unclassified → bounded retries → exhaustion → manual is the safe default for an unconfirmed case — a typed classification invented without contract confirmation would be worse. Happy to reclassify once st0x.deploy confirms the selector's reachability from burn().

File: src/vault/mock.rs (hunk 15 above)
Comment (rouzwelt) [reply src/vault/mock.rs]: The mock documents exactly this limitation at the site ("the fields the stored log carries — it has no to/token"). The four-field contract is pinned where it's real: the service's own find_orchestrator_minted_log_requires_full_match unit tests and the nonce-collision e2e against Anvil. Tightening the mock means adding expected-to/token state that every upstack test constructor then has to thread through, for scenarios those tests already control by nonce/amount.

File: src/vault/mock.rs (hunk 16 above)
Comment (rouzwelt) [reply src/vault/mock.rs:2467]: Verified the claim first: vault-direct prepare_mint_tx only handles PrepareTxFails, so the "same way it corrupts the vault-direct one" framing was false. Doc now states what the test actually verifies: InvalidPreparedMint persists a hash that doesn't match the envelope's bytes, exercising PreparedMintTx::validate rejection. The assertion itself was always real and is unchanged.

File: src/vault/mod.rs (hunk 3 above)
Comment (rouzwelt) [reply src/vault/mod.rs:300]: Production callers land with the orchestrator mint aggregate arms stacked above this commit; the `dead_code` allows on these four methods are removed there, once the callers exist.

File: src/vault/orchestrator.rs (hunk 14 above)
Comment (rouzwelt) [reply src/vault/orchestrator.rs:24]: The struct deliberately mirrors the on-chain MintAuthV1 tuple one-to-one — the permanent event persists exactly what is signed and submitted, and to_binding stays trivial. The mode invariant is enforced at the validation boundary (now including the empty-signature-EOA rejection from this review), which is where an unvalidated authorization is stopped from being stored. An enum would decouple the persisted shape from the contract shape it must reproduce verbatim, and ripples across five stacked branches.

File: src/vault/service.rs (hunk 13 above)
Comment (rouzwelt) [reply src/vault/service.rs:62]: Deliberately deferred: the orchestrator rollout is Base-only, and the window's failure mode is fail-safe (an out-of-window landing drops to the documented manual path, never a false completion). Deriving from chain config is the right move when a second network gets an orchestrator.

File: src/vault/service.rs (hunk 6 above)
Comment (rouzwelt) [reply src/vault/service.rs]: The scan structurally can't detect all near-misses: token and to are topic filters, so a different-token consumption never even reaches the client-side match. The authoritative "consumed" signal is nonceUsed, and the upstack callers already use it: a consumed nonce with no full match escalates to the classified NonceConsumedByOtherMint park (never auto-retried), via the confirm-side recheck. A scan-level distinct error would duplicate that contract with weaker evidence.

File: src/vault/service.rs (hunk 8 above)
Comment (rouzwelt) [reply src/vault/service.rs]: The ordering is already pinned end-to-end: the e2e (orchestrator_mint_end_to_end, upstack) signs over the deployed contract's own mintAuthDigest and completes a real on-chain mint — a transposition here would fail signature verification in that test. A static fixture would pin our call against our own assumption; the e2e pins it against the contract.

File: src/vault/service.rs (hunk 10 above)
Comment (rouzwelt) [reply src/vault/service.rs:1373]: Confirmed direction matters here: normalizing s would accept signatures a low-s-enforcing contract rejects — the wrong direction. As-is, a high-s signature that the orchestrator rejects fails on-chain and classifies (fail-safe). Holding this until st0x.deploy confirms what the contract enforces; if it enforces low-s, the right change is to reject high-s at validation.

File: tests/orchestrator_redemption.rs (hunk 18 above)
Comment (rouzwelt) [reply tests/orchestrator_redemption.rs:171]: Fair on diff hygiene, but this hunk is now merged history: three stacked branches build on the Vec shape (the crash-recovery test and the mint e2e both consume it via slices). Reverting to the array re-opens a conflict cascade through those branches for purely cosmetic gain, so leaving it as-is.

**Comment by rouzwelt** [comment]
@coderabbit review

File: src/vault/service.rs [hunk 19]
```diff
@@ -877,6 +924,323 @@ impl VaultService for RealBlockchainService {
...
+        query: MintedLogQuery,
+        authorization: &MintAuthorization,
+    ) -> Result<(), VaultError> {
+        let MintedLogQuery { orchestrator, to, nonce, token, amount, .. } =
+            query;
+        let contract = IST0xOrchestratorV1::new(orchestrator, &self.provider);
+
+        // Consumed-nonce first: definitive regardless of recipient type, and
+        // a submission with a used nonce could only revert NonceReplayed.
+        if contract.nonceUsed(to, nonce).call().await? {
```
Comment (0xgleb) [inline src/vault/service.rs:1348]: minor: Keep the query and authorization nonce tied together explicitly in `validate_mint_authorization`. Current callers construct `MintedLogQuery.nonce` from `authorization.nonce`, so this is not an exploitable preflight bypass on this head, but an assertion or single nonce source would protect the invariant from future refactors.

File: src/vault/mock.rs [hunk 20]
```diff
@@ -1513,23 +1658,234 @@ impl VaultService for MockVaultService {
...
+            // Mirror the real scan's three-way verdict on the fields the
+            // stored log carries (it has no `to`/`token`): a stored log
+            // whose nonce equals the query's is the pair's one landing —
+            // matching amount is this mint's (`FullMatch`), a differing
+            // amount proves a different mint consumed the pair
+            // (`Mismatch`) — while a missing or different-nonce log means
+            // nothing was found at the pair (`NotFound`).
+            let stored = self.orchestrator.minted_log.lock().unwrap().clone();
+            return Ok(match stored {
+                Some(log) if log.nonce == nonce => {
```
Comment (0xgleb) [inline src/vault/mock.rs:1909]: minor: Align the mock `find_orchestrator_minted_log` matcher with production by checking `to` and `token` as well as nonce and amount. The mismatch can only make recovery tests report a false `FullMatch`; it does not change production behavior, so this is test-double hardening rather than a merge blocker.

**Review by 0xgleb — APPROVED** [review]

**Review by JuaniRios — APPROVED** [review]
